packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,349 @@
+/**
+ * Copyright 2025, SumUp Ltd.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use client';
+
+import {
+  forwardRef,
+  type HTMLAttributes,
+  type ReactNode,
+  type RefObject,
+  useCallback,
+  useEffect,
+  useRef,
+} from 'react';
+
+import type { Locale } from '../../util/i18n.js';
+import { applyMultipleRefs } from '../../util/refs.js';
+import dialogPolyfill from '../../vendor/dialog-polyfill/index.js';
+import { isEscape } from '../../util/key-codes.js';
+import { useScrollLock } from '../../hooks/useScrollLock/useScrollLock.js';
+import { getFirstFocusableElement } from '../Modal/ModalService.js';
+import { CloseButton } from '../CloseButton/index.js';
+import type { ClickEvent } from '../../types/events.js';
+import { clsx } from '../../styles/clsx.js';
+import { useClickOutside } from '../../hooks/useClickOutside/index.js';
+import { useEscapeKey } from '../../hooks/useEscapeKey/index.js';
+import { useLatest } from '../../hooks/useLatest/index.js';
+
+import classes from './dialog.module.css';
+
+export const ANIMATION_DURATION = 300;
+
+type DataAttribute = `data-${string}`;
+
+export interface DialogProps
+  extends Omit<HTMLAttributes<HTMLDialogElement>, 'children'> {
+  /**
+   * Whether the modal dialog is open or not.
+   */
+  open: boolean;
+  /**
+   * Whether the dialog is modal or not.
```

Comment:
This is a good opportunity to educate users about the underlying native `dialog` API. Let's link to the `dialog` docs for developers who are curious about the difference (e.g. https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal).